brdnet/classifier.py

```diff
@@ -96,6 +102,103 @@ def reduce_dimensionality(expression_df, Z_df):
     return reduced_matrix
 
 
+def get_validation_set(data, validation_fraction=.2):
+    '''Split a dataframe into training and validation data by extracting studies
+       that contain a certain fraction of the samples
+
+       Arguments
+       ---------
+       data: pandas.DataFrame
+           data is a dataframe where the rows represent genes and columns represent samples.
+           The column names should be of the format 'studyid.runid' to allow the study
+           information to be used.
+       validation_fraction: float
+           The fraction of the dataset to be pulled out as validation data
+
+       Returns
+       -------
+       train_df: pandas.DataFrame
+           A dataframe containing the fraction of the sample to be used in training
+       val_df:
+           A dataframe containing the fraction of the sample to be used for validation
+    '''
+    studies = [name.split('.')[0] for name in data.columns]
+    # random.shuffle shuffles in place
+    random.shuffle(studies)
+
+    counter = Counter(studies)
+
+    target_sample_count = len(data.columns) * validation_fraction
+
+    samples_so_far = 0
+    val_studies = []
+    for study, samples in counter.items():
+        # Prevent the validation set from being too much larger than the target
+        if samples_so_far + samples > len(data.columns) * (validation_fraction + .05):
+            continue
+        val_studies.append(study)
+        samples_so_far += samples
+
+        if samples_so_far >= target_sample_count:
+            break
+
+    val_columns = []
+    for study in val_studies:
+        val_columns.extend([col for col in data.columns if study in col])
+
+    val_df = data[val_columns]
+    train_df = data[data.columns.difference(val_columns)]
+
+    return train_df, val_df
+
+
+def prepare_input_data(Z_df, healthy_df, disease_df):
+    '''Convert the dataframes from run_plier and download_categorized_data into
+    training and validation datasets with accompanying labels
+
+    Arguments
+    ---------
+    Z_df: pandas.DataFrame
+        The matrix to convert the expression data into the PLIER latent space
+    healthy_df:
+        The dataframe containing healthy gene expression samples
+    disease_df:
+        The dataframe containin unhealthy gene expression samples
+
+    Returns
+    -------
+    train_X: numpy.array
+        A numpy array containing the training gene expression data
+    train_Y: numpy.array
+        The labels corresponding to whether each sample represents healthy or unhealthy
+        gene expression
+    val_X: numpy.array
+        The gene expression data to be held out to evaluate model training
+    val_Y: numpy.array
+        The labels for val_X
+    '''
+    healthy_train, healthy_val = get_validation_set(healthy_df)
+    healthy_train = reduce_dimensionality(healthy_train, Z_df)
+    healthy_val = reduce_dimensionality(healthy_val, Z_df)
+
+    disease_train, disease_val = get_validation_set(disease_df)
```

Comment:
Do you want to have the same partitions for both the healthy and disease datasets? So, for instance, if study 1 and 2 are the training set and study 3 is the validation set for the healthy samples, do you want to have the same 1+2 / 3 partition for the disease samples?

It seems like the way you've written this, you could get different partitions for healthy vs. disease (but correct me if I'm wrong, or if this is intentional).

Reply:
Yep, I totally forgot to synchronize the healthy and unhealthy studies. I modified the function to fix this.